File: crypto/src/test/cplusplus/z2.cpp
/*
 * Copyright (c) 2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <boost/test/unit_test.hpp>

#include "z2.h"

using namespace blacknet::crypto;

BOOST_AUTO_TEST_SUITE(Z2s)

BOOST_AUTO_TEST_CASE(Representative) {
    Z2 a(-1);
    BOOST_TEST(1 == a.canonical());
    BOOST_TEST(1 == a.balanced());
    BOOST_TEST(1 == a.absolute());
}

BOOST_AUTO_TEST_CASE(Add) {
    BOOST_TEST(Z2(0) == Z2(0) + Z2(0));
    BOOST_TEST(Z2(1) == Z2(0) + Z2(1));
    BOOST_TEST(Z2(1) == Z2(1) + Z2(0));
    BOOST_TEST(Z2(0) == Z2(1) + Z2(1));
}

BOOST_AUTO_TEST_CASE(Mul) {
    BOOST_TEST(Z2(0) == Z2(0) * Z2(0));
    BOOST_TEST(Z2(0) == Z2(0) * Z2(1));
    BOOST_TEST(Z2(0) == Z2(1) * Z2(0));
    BOOST_TEST(Z2(1) == Z2(1) * Z2(1));
}

BOOST_AUTO_TEST_CASE(Sqr) {
    BOOST_TEST(Z2(0) == Z2(0).square());
    BOOST_TEST(Z2(1) == Z2(1).square());
}

BOOST_AUTO_TEST_CASE(Sub) {
    BOOST_TEST(Z2(0) == Z2(0) - Z2(0));
    BOOST_TEST(Z2(1) == Z2(0) - Z2(1));
    BOOST_TEST(Z2(1) == Z2(1) - Z2(0));
    BOOST_TEST(Z2(0) == Z2(1) - Z2(1));
}

BOOST_AUTO_TEST_CASE(Inv) {
    BOOST_TEST(Z2(1) == Z2(1).invert().value());
    BOOST_TEST(!Z2(0).invert());
}

BOOST_AUTO_TEST_CASE(InfinityNorm) {
    BOOST_TEST(!Z2(0).checkInfinityNorm(0));
    BOOST_TEST(Z2(0).checkInfinityNorm(1));
    BOOST_TEST(!Z2(1).checkInfinityNorm(0));
    BOOST_TEST(!Z2(1).checkInfinityNorm(1));
}

BOOST_AUTO_TEST_SUITE_END()

File: crypto/src/test/rust/z2.rs
/*
 * Copyright (c) 2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_crypto::norm::InfinityNorm;
use blacknet_crypto::operation::{Inv, Square};
use blacknet_crypto::ring::IntegerRing;
use blacknet_crypto::z2::Z2;

#[test]
fn congruence() {
    assert_eq!(Z2::new(-2), Z2::new(2));
    assert_ne!(Z2::new(-3), Z2::new(2));
}

#[test]
fn add() {
    assert_eq!(Z2::new(0) + Z2::new(0), Z2::new(0));
    assert_eq!(Z2::new(0) + Z2::new(1), Z2::new(1));
    assert_eq!(Z2::new(1) + Z2::new(0), Z2::new(1));
    assert_eq!(Z2::new(1) + Z2::new(1), Z2::new(0));
}

#[test]
fn mul() {
    assert_eq!(Z2::new(0) * Z2::new(0), Z2::new(0));
    assert_eq!(Z2::new(0) * Z2::new(1), Z2::new(0));
    assert_eq!(Z2::new(1) * Z2::new(0), Z2::new(0));
    assert_eq!(Z2::new(1) * Z2::new(1), Z2::new(1));
}

#[test]
fn sqr() {
    assert_eq!(Z2::new(0).square(), Z2::new(0));
    assert_eq!(Z2::new(1).square(), Z2::new(1));
}

#[test]
fn sub() {
    assert_eq!(Z2::new(0) - Z2::new(0), Z2::new(0));
    assert_eq!(Z2::new(0) - Z2::new(1), Z2::new(1));
    assert_eq!(Z2::new(1) - Z2::new(0), Z2::new(1));
    assert_eq!(Z2::new(1) - Z2::new(1), Z2::new(0));
}

#[test]
fn inv() {
    assert_eq!(Z2::new(1).inv().unwrap(), Z2::new(1));
    assert!(Z2::new(0).inv().is_none());
}

#[test]
fn infinity_norm() {
    assert!(!Z2::new(0).check_infinity_norm(0));
    assert!(Z2::new(0).check_infinity_norm(1));
    assert!(!Z2::new(1).check_infinity_norm(0));
    assert!(!Z2::new(1).check_infinity_norm(1));
}
